File: radix_tree.py


class Node:
    def __init__(self):
        self.children = [None] * 10
        self.is_end_of_SERIE = False
        self.dict_of_ranges = {}

class Radix_tree:
    def __init__(self):
        self.root = Node()
    
    def insert(self, numeracion_node): ##Important NIR, SERIE, Inicial - Numeracion Final
        ####First organize first six values of phone number into three
        NIR = numeracion_node.NIR
        SERIE = numeracion_node.SERIE
        first_six_numbers = NIR + SERIE
        curr_node = self.root
        for num in first_six_numbers:
            if curr_node[int(num)] is None:
                curr_node[int(num)] = Node()
            curr_node = curr_node(int(num))
        ###Now organize values in range values, add to list and set list as 
        ##Key value in leaf node with dictionary
        initial_value = int(numeracion_node.NUM_INICIAL)
        final_value = int(numeracion_node.NUM_FINAL)
        list_of_range = []
        list_of_range[0] = initial_value
        list_of_range[1] = final_value
        curr_node.is_end_of_SERIE = True
        curr_node.dict_of_ranges[list_of_range] = numeracion_node


    def search(self, num):
        if (len(num) != 10):
            return "You did not enter a valid phone number"
        if (num.isdigit() == False):
            return "You did not enter a valid phone number"
        first_six = num[0] + num[1] + num[2] + num[3] + num[4] + num[5]
        last_four = num[6] + num[7] + num[8] + num[9]
        curr_node = self.root
        for path in first_six:
            if curr_node[int(path)] == None:
                return "Phone number not in database"
            curr_node = curr_node[int(path)]
        if (curr_node.is_end_of_SERIE == False):
            return "Something wrong with tree"
        for keys in curr_node.dict_of_ranges:
            initial = int(keys[0])
            final = int(keys[1])
            if initial <= last_four <= final:
                numeracion = curr_node.dict_of_ranges[keys]
                return f"Razon social: {numeracion.razon_social}, Tipo Red: {numeracion.tipo_red}"

File: JavaScript Trie/radix_tree.js
//npm run dev to run React site

class Node{
    constructor(){
        this.children = {};
        this.is_END_of_SERIE = false;
        this.info = [];
    }
}

class EndNode{
    constructor(start_val, end_val, info){
        this.start_val = start_val;
        this.end_val = end_val;
        this.information = info;
    }
}

class Radix_tree{
    constructor(){
        this.root = new Node();
    }

    insert(numeracion_node) {
        let NIR = numeracion_node.NIR;
        let SERIE = numeracion_node.SERIE;
        let first_six_numbers = NIR + SERIE;
        let curr_node = this.root;
        for (const num in first_six_numbers){
            if (!curr_node.children[num]){
                curr_node.children[num] = new Node();
            }
            curr_node = curr_node.children[num];
        }
        let start_val = Number(numeracion_node.NUM_INICIAL);
        let end_val = Number(numeracion_node.NUM_FINAL);
        let information = numeracion_node;
        let newEndNode = new EndNode(start_val, end_val, information);
        curr_node.info.push(newEndNode);
    }

    search(num) {
        if (num.length != 10){
            return "You did not enter a valid phone number";
        }
        if (isNaN(num)){
            return "You did not enter a valid phone number";
        }
        let first_six = num.slice(0, 6);
        let last_four = num.slice(6, 10);
        curr_node = self.root;
        for (val in first_six){
            if (!curr_node.children[val]){
                return "Phone Number not in Database";
            }
            curr_node = curr_node.children[val];
        }
        if (curr_node.is_END_of_SERIE == false){
            return "Something wrong with the tree";
        }
        for (childNode in curr_node.info){
            if (childNode.start_val <= Number(last_four) && Number(last_four) <= childNode.end_val){
                let numeracion = childNode.information;
                return `Razon social: ${numeracion.razon_social}, Tipo Red: ${numeracion.tipo_red}, Poblacion: ${numeracion.poblacion}`
            }
        }
        return "Phone number not in Database"
    }
}

export default Radix_tree